src/main/java/com/google/devtools/build/lib/remote/RemoteExternalOverlayFileSystem.java

```diff
@@ -179,31 +180,60 @@ public void afterCommand() {
     this.evaluator = null;
   }
 
-  public void injectRemoteRepo(RepositoryName repo, Tree remoteContents, String markerFile)
-      throws IOException {
+  /**
+   * Injects the given remote contents, possibly prefetching some files, and returns true on
+   * success.
+   */
+  public boolean injectRemoteRepo(RepositoryName repo, Tree remoteContents, String markerFile)
+      throws IOException, InterruptedException {
     var childMap =
         remoteContents.getChildrenList().stream()
             .collect(
                 toImmutableMap(cache.digestUtil::compute, directory -> directory, (a, b) -> a));
+    var repoDir = externalDirectory.getChild(repo.getName());
+    var filesToPrefetch = new ArrayList<PathFragment>();
     injectRecursively(
-        externalFs, externalDirectory.getChild(repo.getName()), remoteContents.getRoot(), childMap);
+        externalFs,
+        repoDir,
+        remoteContents.getRoot(),
+        childMap,
+        path -> {
```

Comment:
maybe just inline this argument (as in, replace it with `filesToPrefetch`)? or do you expect to have other ways of collecting paths?